src/main/host/descriptor/timerfd.rs

```diff
@@ -0,0 +1,275 @@
+        false
+    }
+
+    pub fn set_has_open_file(&mut self, val: bool) {
+        self.has_open_file = val;
+    }
+
+    pub fn readv(
+        &mut self,
+        iovs: &[IoVec],
+        offset: Option<libc::off_t>,
+        _flags: libc::c_int,
+        mem: &mut MemoryManager,
+        cb_queue: &mut CallbackQueue,
+    ) -> Result<libc::ssize_t, SyscallError> {
+        // TimerFds don't support seeking
+        if offset.is_some() {
+            return Err(Errno::ESPIPE.into());
+        }
+
+        // timerfd_create(2): "read(2) returns an unsigned 8-byte integer containing the number of
+        // expirations that have occurred."
+        const NUM_BYTES: usize = 8;
+
+        let len: libc::size_t = iovs.iter().map(|x| x.len).sum();
+
+        // This check doesn't guarantee that we can write all bytes since the stream length is only
+        // a hint.
+        if len < NUM_BYTES {
+            log::trace!(
+                "Reading from TimerFd requires a buffer of at least {} bytes",
+                NUM_BYTES
+            );
+            return Err(Errno::EINVAL.into());
+        }
+
+        let expiration_count = self.timer.consume_expiration_count();
+
+        if expiration_count == 0 {
+            log::trace!("TimerFd expiration count is 0 and cannot be read right now");
+            return Err(Errno::EWOULDBLOCK.into());
+        }
+
+        let mut writer = IoVecWriter::new(iovs, mem);
+        let to_write: [u8; NUM_BYTES] = expiration_count.to_ne_bytes();
+        writer.write_all(&to_write)?;
+
+        // We just read the expiration counter and so are not readable anymore.
+        self.update_state(cb_queue);
+
+        Ok(NUM_BYTES.try_into().unwrap())
+    }
+
+    pub fn writev(
+        &mut self,
+        _iovs: &[IoVec],
+        _offset: Option<libc::off_t>,
+        _flags: libc::c_int,
+        _mem: &mut MemoryManager,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<libc::ssize_t, SyscallError> {
+        // TimerFds don't support writing.
+        Err(Errno::EINVAL.into())
+    }
+
+    pub fn close(&mut self, cb_queue: &mut CallbackQueue) -> Result<(), SyscallError> {
+        // Set the closed flag and remove the active and readable flags.
+        self.copy_state(
+            FileState::CLOSED | FileState::ACTIVE | FileState::READABLE,
+            FileState::CLOSED,
+            cb_queue,
+        );
+
+        Ok(())
+    }
+
+    pub fn ioctl(
+        &mut self,
+        request: u64,
+        _arg_ptr: ForeignPtr<()>,
+        _memory_manager: &mut MemoryManager,
+    ) -> SyscallResult {
+        // The only timerfd-specific ioctl request is for `TFD_IOC_SET_TICKS`, which is available
+        // since Linux 3.17 but only if the kernel was configured with `CONFIG_CHECKPOINT_RESTORE`.
+        // See timerfd_create(2) for more details.
+        log::warn!("We do not yet handle ioctl request {} on TimerFds", request);
+        Err(Errno::EINVAL.into())
+    }
```

Comment:
Also, it would be nicer here to return `ENOTSUP` when the ioctl request is `TFD_IOC_SET_TICKS ` and then `EINVAL`  otherwise. However, I'm not finding the `TFD_IOC_SET_TICKS` constant in the libc crate.

Reply:
Yeah, I'd be more comfortable with just logging at debug level if we returned something like `ENOTSUP`, since `EINVAL` incorrectly indicates "you passed bad arguments". OTOH `ENOTSUP` isn't actually listed as a possible error in the `ioctl` man page, so that's not really "right" either.

I like the idea of returning `EINVAL` and logging a warning once... ideally for each distinct value of `request` that we don't support.

Maybe keep it as `warn` (and `EINVAL`) for now and add a `TODO` to make it warn-once when we implement that function?

Reply:
I can see about exporting `TFD_IOC_SET_TICKS` (and `itimerspec` etc) in the `linux-api` crate.

Reply:
Okay sounds good to me. I think it can be merged as-is, and we can update it later if we want.